File: app/typescript-2/src/tests/player.test.ts
import { describe, it, expect, beforeEach } from 'vitest'
import { Player } from '../player'
import { Config } from '../config'
import { Stage } from '../stage'
import { PuyoImage } from '../puyoimage'

describe('プレイヤー', () => {
  let config: Config
  let puyoImage: PuyoImage
  let stage: Stage
  let player: Player

  beforeEach(() => {
    // DOMの準備
    document.body.innerHTML = `
      <div id="stage"></div>
    `
    config = new Config()
    puyoImage = new PuyoImage(config)
    stage = new Stage(config, puyoImage)
    player = new Player(config, stage, puyoImage)
  })

  describe('キー入力', () => {
    it('左キーが押されると、左向きの移動フラグが立つ', () => {
      // キーダウンイベントをシミュレート（左キー）
      const event = new KeyboardEvent('keydown', { key: 'ArrowLeft' })
      document.dispatchEvent(event)

      expect(player['inputKeyLeft']).toBe(true)
    })

    it('右キーが押されると、右向きの移動フラグが立つ', () => {
      // キーダウンイベントをシミュレート（右キー）
      const event = new KeyboardEvent('keydown', { key: 'ArrowRight' })
      document.dispatchEvent(event)

      expect(player['inputKeyRight']).toBe(true)
    })

    it('キーが離されると、対応する移動フラグが下がる', () => {
      // まず左キーを押す
      document.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowLeft' }))
      expect(player['inputKeyLeft']).toBe(true)

      // 次に左キーを離す
      document.dispatchEvent(new KeyboardEvent('keyup', { key: 'ArrowLeft' }))
      expect(player['inputKeyLeft']).toBe(false)
    })
  })

  describe('ぷよの移動', () => {
    beforeEach(() => {
      // 新しいぷよを作成
      player.createNewPuyo()
    })

    it('左に移動できる場合、左に移動する', () => {
      // 初期位置を記録
      const initialX = player['puyoX']

      // 左に移動
      player.moveLeft()

      // 位置が1つ左に移動していることを確認
      expect(player['puyoX']).toBe(initialX - 1)
    })

    it('右に移動できる場合、右に移動する', () => {
      // 初期位置を記録
      const initialX = player['puyoX']

      // 右に移動
      player.moveRight()

      // 位置が1つ右に移動していることを確認
      expect(player['puyoX']).toBe(initialX + 1)
    })

    it('左端にいる場合、左に移動できない', () => {
      // 左端に移動
      player['puyoX'] = 0

      // 左に移動を試みる
      player.moveLeft()

      // 位置が変わっていないことを確認
      expect(player['puyoX']).toBe(0)
    })

    it('右端にいる場合、右に移動できない', () => {
      // 右端に移動（ステージの幅 - 1）
      player['puyoX'] = config.stageCols - 1

      // 右に移動を試みる
      player.moveRight()

      // 位置が変わっていないことを確認
      expect(player['puyoX']).toBe(config.stageCols - 1)
    })
  })

  describe('ぷよの回転', () => {
    beforeEach(() => {
      // 新しいぷよを作成
      player.createNewPuyo()
    })

    it('時計回りに回転すると、回転状態が1増える', () => {
      // 初期回転状態を記録
      const initialRotation = player['rotation']

      // 時計回りに回転
      player.rotateRight()

      // 回転状態が1増えていることを確認
      expect(player['rotation']).toBe((initialRotation + 1) % 4)
    })

    it('反時計回りに回転すると、回転状態が1減る', () => {
      // 初期回転状態を記録
      const initialRotation = player['rotation']

      // 反時計回りに回転
      player.rotateLeft()

      // 回転状態が1減っていることを確認（負の値にならないように調整）
      expect(player['rotation']).toBe((initialRotation + 3) % 4)
    })

    it('回転状態が4になると0に戻る', () => {
      // 回転状態を3に設定
      player['rotation'] = 3

      // 時計回りに回転
      player.rotateRight()

      // 回転状態が0になっていることを確認
      expect(player['rotation']).toBe(0)
    })
  })

  describe('壁キック処理', () => {
    beforeEach(() => {
      // 新しいぷよを作成
      player.createNewPuyo()
    })

    it('右端で右回転すると、左に移動して回転する（壁キック）', () => {
      // 右端に移動
      player['puyoX'] = config.stageCols - 1
      player['rotation'] = 0 // 上向き

      // 右回転（2つ目のぷよが右にくる）
      player.rotateRight()

      // 壁キックにより左に移動していることを確認
      expect(player['puyoX']).toBe(config.stageCols - 2)
      expect(player['rotation']).toBe(1)
    })

    it('左端で左回転すると、右に移動して回転する（壁キック）', () => {
      // 左端に移動
      player['puyoX'] = 0
      player['rotation'] = 0 // 上向き

      // 左回転（2つ目のぷよが左にくる）
      player.rotateLeft()

      // 壁キックにより右に移動していることを確認
      expect(player['puyoX']).toBe(1)
      expect(player['rotation']).toBe(3)
    })
  })

  describe('自由落下', () => {
    beforeEach(() => {
      // 新しいぷよを作成
      player.createNewPuyo()
    })

    it('指定時間が経過すると、ぷよが1マス下に落ちる', () => {
      // 初期位置を記録
      const initialY = player['puyoY']

      // 落下間隔を取得（例: 1000ms = 1秒）
      const dropInterval = 1000

      // ゲームの更新処理を実行（落下間隔分）
      player.update(dropInterval)

      // 位置が1つ下に移動していることを確認
      expect(player['puyoY']).toBe(initialY + 1)
    })

    it('指定時間未満では、ぷよは落ちない', () => {
      // 初期位置を記録
      const initialY = player['puyoY']

      // 落下間隔を取得
      const dropInterval = 1000

      // タイマーを半分だけ進める
      player.update(dropInterval / 2)

      // 位置が変わっていないことを確認
      expect(player['puyoY']).toBe(initialY)
    })

    it('下端に達した場合、それ以上落ちない', () => {
      // 下端の1つ上に配置
      player['puyoY'] = config.stageRows - 1

      // 落下処理を実行
      player.update(1000)

      // 位置が変わっていないことを確認（下端を超えない）
      expect(player['puyoY']).toBe(config.stageRows - 1)
    })
  })

  describe('着地処理', () => {
    it('下端に着地すると、着地フラグが立つ', () => {
      // 下端に配置（子ぷよが上にあるので、親ぷよはstageRows - 1）
      player['puyoY'] = config.stageRows - 1

      // 落下処理を実行（これ以上落ちられないので着地）
      player.update(1000)

      // 着地フラグが立っていることを確認
      expect(player.isLanded()).toBe(true)
    })

    it('着地していない状態では、着地フラグが立たない', () => {
      // 初期位置（上の方）
      player.createNewPuyo()

      // 着地フラグが立っていないことを確認
      expect(player.isLanded()).toBe(false)
    })

    it('他のぷよの上に着地すると、着地フラグが立つ', () => {
      // ステージに既存のぷよを配置（下端）
      stage.setPuyo(2, config.stageRows - 1, 1)

      // その上に新しいぷよを配置
      player.createNewPuyo()
      player['puyoX'] = 2
      player['puyoY'] = config.stageRows - 2

      // 落下処理を実行（既存のぷよの上で着地）
      player.update(1000)

      // 着地フラグが立っていることを確認
      expect(player.isLanded()).toBe(true)
    })

    it('他のぷよがある場所には移動できない', () => {
      // ステージに既存のぷよを配置
      stage.setPuyo(2, 5, 1)

      // その真上に新しいぷよを配置
      player.createNewPuyo()
      player['puyoX'] = 2
      player['puyoY'] = 3

      // 落下処理を実行（既存のぷよまで落ちる）
      player.update(1000)
      player.update(1000)

      // 既存のぷよの上で止まっていることを確認
      expect(player['puyoY']).toBe(4)
      expect(player.isLanded()).toBe(true)
    })
  })

  describe('回転時の衝突判定', () => {
    it('他のぷよがある場所には回転できない', () => {
      // ステージに既存のぷよを配置
      stage.setPuyo(4, 5, 1) // 右隣にぷよを配置

      // その隣に新しいぷよを配置（子ぷよは上）
      player.createNewPuyo()
      player['puyoX'] = 3
      player['puyoY'] = 5
      player['rotation'] = 0 // 子ぷよが上

      // 回転状態を記録
      const initialRotation = player['rotation']

      // 右回転を試みる（子ぷよが右に来る = 既存のぷよと重なる）
      player.rotateRight()

      // 回転できず、回転状態が変わっていないことを確認
      expect(player['rotation']).toBe(initialRotation)
    })

    it('回転可能な場合は回転できる', () => {
      // 周囲に何もない位置に配置
      player.createNewPuyo()
      player['puyoX'] = 3
      player['puyoY'] = 5
      player['rotation'] = 0

      // 右回転
      player.rotateRight()

      // 回転できていることを確認
      expect(player['rotation']).toBe(1)
    })
  })

  describe('高速落下', () => {
    beforeEach(() => {
      // 新しいぷよを作成
      player.createNewPuyo()
    })

    it('下キーが押されていると、落下速度が上がる', () => {
      // 下キーを押す
      document.dispatchEvent(new KeyboardEvent('keydown', { key: 'ArrowDown' }))

      // 通常の落下処理
      const normalDropSpeed = 1
      const fastDropSpeed = player.getDropSpeed()

      // 高速落下の速度が通常より速いことを確認
      expect(fastDropSpeed).toBeGreaterThan(normalDropSpeed)
    })

    it('下に移動できる場合、下に移動する', () => {
      // 初期位置を記録
      const initialY = player['puyoY']

      // 下に移動
      player.moveDown()

      // 位置が1つ下に移動していることを確認
      expect(player['puyoY']).toBe(initialY + 1)
    })

    it('下に障害物がある場合、下に移動できない', () => {
      // ステージの一番下に移動
      player['puyoY'] = config.stageRows - 1

      // 下に移動を試みる
      const canMove = player.moveDown()

      // 移動できないことを確認
      expect(canMove).toBe(false)
      expect(player['puyoY']).toBe(config.stageRows - 1)
    })
  })
})

File: app/csharp-2/blazor/PuyoPuyoTDD.Tests/Models/PlayerTest.cs
// <copyright file="PlayerTest.cs" company="PuyoPuyoTDD">
// Copyright (c) PuyoPuyoTDD. All rights reserved.
// </copyright>

using PuyoPuyoTDD.Models;
using Xunit;

namespace PuyoPuyoTDD.Tests.Models;

/// <summary>
/// Player クラスのテスト.
/// </summary>
public class PlayerTest
{
    private readonly Config config;
    private readonly Stage stage;
    private readonly Player player;

    /// <summary>
    /// Initializes a new instance of the <see cref="PlayerTest"/> class.
    /// </summary>
    public PlayerTest()
    {
        this.config = new Config();
        this.stage = new Stage(this.config);
        this.player = new Player(this.config, this.stage);
    }

    /// <summary>
    /// 左キーが押されると左向きの移動フラグが立つかテスト.
    /// </summary>
    [Fact]
    public void 左キーが押されると左向きの移動フラグが立つ()
    {
        // Act
        this.player.SetInputLeft(true);

        // Assert
        Assert.True(this.player.InputKeyLeft);
    }

    /// <summary>
    /// 右キーが押されると右向きの移動フラグが立つかテスト.
    /// </summary>
    [Fact]
    public void 右キーが押されると右向きの移動フラグが立つ()
    {
        // Act
        this.player.SetInputRight(true);

        // Assert
        Assert.True(this.player.InputKeyRight);
    }

    /// <summary>
    /// キーが離されると対応する移動フラグが下がるかテスト.
    /// </summary>
    [Fact]
    public void キーが離されると対応する移動フラグが下がる()
    {
        // Arrange
        this.player.SetInputLeft(true);
        Assert.True(this.player.InputKeyLeft);

        // Act
        this.player.SetInputLeft(false);

        // Assert
        Assert.False(this.player.InputKeyLeft);
    }

    /// <summary>
    /// 左に移動できる場合左に移動するかテスト.
    /// </summary>
    [Fact]
    public void 左に移動できる場合左に移動する()
    {
        // Arrange
        this.player.CreateNewPuyo();
        var initialX = this.player.PuyoX;

        // Act
        this.player.MoveLeft();

        // Assert
        Assert.Equal(initialX - 1, this.player.PuyoX);
    }

    /// <summary>
    /// 右に移動できる場合右に移動するかテスト.
    /// </summary>
    [Fact]
    public void 右に移動できる場合右に移動する()
    {
        // Arrange
        this.player.CreateNewPuyo();
        var initialX = this.player.PuyoX;

        // Act
        this.player.MoveRight();

        // Assert
        Assert.Equal(initialX + 1, this.player.PuyoX);
    }

    /// <summary>
    /// 左端にいる場合左に移動できないかテスト.
    /// </summary>
    [Fact]
    public void 左端にいる場合左に移動できない()
    {
        // Arrange
        this.player.CreateNewPuyo();
        this.player.SetPuyoX(0);

        // Act
        this.player.MoveLeft();

        // Assert
        Assert.Equal(0, this.player.PuyoX);
    }

    /// <summary>
    /// 右端にいる場合右に移動できないかテスト.
    /// </summary>
    [Fact]
    public void 右端にいる場合右に移動できない()
    {
        // Arrange
        this.player.CreateNewPuyo();
        this.player.SetPuyoX(this.config.StageWidth - 1);

        // Act
        this.player.MoveRight();

        // Assert
        Assert.Equal(this.config.StageWidth - 1, this.player.PuyoX);
    }

    /// <summary>
    /// 時計回りに回転すると回転状態が1増えるかテスト.
    /// </summary>
    [Fact]
    public void 時計回りに回転すると回転状態が1増える()
    {
        // Arrange
        this.stage.Initialize();
        this.player.CreateNewPuyo();
        this.player.SetPuyoY(5); // 安全な位置に配置
        var initialRotation = this.player.Rotation;

        // Act
        this.player.RotateRight();

        // Assert
        Assert.Equal((initialRotation + 1) % 4, this.player.Rotation);
    }

    /// <summary>
    /// 反時計回りに回転すると回転状態が1減るかテスト.
    /// </summary>
    [Fact]
    public void 反時計回りに回転すると回転状態が1減る()
    {
        // Arrange
        this.stage.Initialize();
        this.player.CreateNewPuyo();
        this.player.SetPuyoY(5); // 安全な位置に配置
        var initialRotation = this.player.Rotation;

        // Act
        this.player.RotateLeft();

        // Assert
        Assert.Equal((initialRotation + 3) % 4, this.player.Rotation);
    }

    /// <summary>
    /// 回転状態が4になると0に戻るかテスト.
    /// </summary>
    [Fact]
    public void 回転状態が4になると0に戻る()
    {
        // Arrange
        this.stage.Initialize();
        this.player.CreateNewPuyo();
        this.player.SetPuyoY(5); // 安全な位置に配置
        this.player.SetRotation(3);

        // Act
        this.player.RotateRight();

        // Assert
        Assert.Equal(0, this.player.Rotation);
    }

    /// <summary>
    /// 下キーが押されていると落下速度が上がるかテスト.
    /// </summary>
    [Fact]
    public void 下キーが押されていると落下速度が上がる()
    {
        // Arrange
        this.player.SetInputDown(true);

        // Act
        var dropSpeed = this.player.GetDropSpeed();

        // Assert
        Assert.True(dropSpeed > 1);
    }

    /// <summary>
    /// 下に移動できる場合下に移動するかテスト.
    /// </summary>
    [Fact]
    public void 下に移動できる場合下に移動する()
    {
        // Arrange
        this.player.CreateNewPuyo();
        var initialY = this.player.PuyoY;

        // Act
        var canMove = this.player.MoveDown();

        // Assert
        Assert.True(canMove);
        Assert.Equal(initialY + 1, this.player.PuyoY);
    }

    /// <summary>
    /// 下に障害物がある場合下に移動できないかテスト.
    /// </summary>
    [Fact]
    public void 下に障害物がある場合下に移動できない()
    {
        // Arrange
        this.player.CreateNewPuyo();
        this.player.SetPuyoY(this.config.StageHeight - 1);

        // Act
        var canMove = this.player.MoveDown();

        // Assert
        Assert.False(canMove);
        Assert.Equal(this.config.StageHeight - 1, this.player.PuyoY);
    }

    /// <summary>
    /// 回転が下向き(2)の場合子ぷよが地面にめり込まないかテスト.
    /// </summary>
    [Fact]
    public void 回転が下向きの場合子ぷよが地面にめり込まない()
    {
        // Arrange
        this.player.CreateNewPuyo();
        this.player.SetRotation(2); // 下向き
        this.player.SetPuyoY(this.config.StageHeight - 2);

        // Act
        var canMove = this.player.MoveDown();

        // Assert
        Assert.False(canMove);
        Assert.Equal(this.config.StageHeight - 2, this.player.PuyoY);
    }

    /// <summary>
    /// 回転が上向き(0)の場合は最下段まで移動できるかテスト.
    /// </summary>
    [Fact]
    public void 回転が上向きの場合は最下段まで移動できる()
    {
        // Arrange
        this.player.CreateNewPuyo();
        this.player.SetRotation(0); // 上向き
        this.player.SetPuyoY(this.config.StageHeight - 2);

        // Act
        var canMove = this.player.MoveDown();

        // Assert
        Assert.True(canMove);
        Assert.Equal(this.config.StageHeight - 1, this.player.PuyoY);
    }

    /// <summary>
    /// ぷよが着地したかどうかを判定できるかテスト.
    /// </summary>
    [Fact]
    public void ぷよが着地したかどうかを判定できる()
    {
        // Arrange
        this.player.CreateNewPuyo();
        this.player.SetPuyoY(this.config.StageHeight - 1);

        // Act
        var hasLanded = this.player.HasLanded();

        // Assert
        Assert.True(hasLanded);
    }

    /// <summary>
    /// ぷよが空中にいる場合は着地していないと判定されるかテスト.
    /// </summary>
    [Fact]
    public void ぷよが空中にいる場合は着地していないと判定される()
    {
        // Arrange
        this.player.CreateNewPuyo();
        this.player.SetPuyoY(5);

        // Act
        var hasLanded = this.player.HasLanded();

        // Assert
        Assert.False(hasLanded);
    }

    /// <summary>
    /// 着地したぷよをステージに配置できるかテスト.
    /// </summary>
    [Fact]
    public void 着地したぷよをステージに配置できる()
    {
        // Arrange
        this.stage.Initialize();
        this.player.CreateNewPuyo();
        this.player.SetPuyoX(1);
        this.player.SetPuyoY(11);
        this.player.SetRotation(0); // 上向き
        var puyoType = this.player.PuyoType;

        // Act
        this.player.PlacePuyoOnStage();

        // Assert
        // 軸ぷよ (1, 11) と子ぷよ (1, 10) がステージに配置される
        Assert.Equal(puyoType, this.stage.GetPuyo(1, 11));
        Assert.Equal(puyoType, this.stage.GetPuyo(1, 10));
    }

    /// <summary>
    /// ステージのぷよと衝突する場合は下に移動できないかテスト.
    /// </summary>
    [Fact]
    public void ステージのぷよと衝突する場合は下に移動できない()
    {
        // Arrange
        this.stage.Initialize();
        this.stage.SetPuyo(2, 10, 1); // ステージにぷよを配置
        this.player.CreateNewPuyo();
        this.player.SetPuyoX(2);
        this.player.SetPuyoY(9);
        this.player.SetRotation(0); // 上向き

        // Act
        var canMove = this.player.MoveDown();

        // Assert
        Assert.False(canMove);
    }

    /// <summary>
    /// 子ぷよがステージのぷよと衝突する場合は下に移動できないかテスト.
    /// </summary>
    [Fact]
    public void 子ぷよがステージのぷよと衝突する場合は下に移動できない()
    {
        // Arrange
        this.stage.Initialize();
        this.stage.SetPuyo(2, 9, 1); // ステージにぷよを配置
        this.player.CreateNewPuyo();
        this.player.SetPuyoX(2);
        this.player.SetPuyoY(9);
        this.player.SetRotation(0); // 上向き（子ぷよが y=8 にある）

        // Act
        var canMove = this.player.MoveDown();

        // Assert
        // 下に移動すると子ぷよが (2, 9) に衝突するので移動できない
        Assert.False(canMove);
    }

    /// <summary>
    /// ステージのぷよと衝突する場合は左に移動できないかテスト.
    /// </summary>
    [Fact]
    public void ステージのぷよと衝突する場合は左に移動できない()
    {
        // Arrange
        this.stage.Initialize();
        this.stage.SetPuyo(1, 5, 1); // ステージにぷよを配置
        this.player.CreateNewPuyo();
        this.player.SetPuyoX(2);
        this.player.SetPuyoY(5);
        this.player.SetRotation(0); // 上向き

        // Act
        this.player.MoveLeft();

        // Assert
        Assert.Equal(2, this.player.PuyoX); // 移動できないので位置は変わらない
    }

    /// <summary>
    /// ステージのぷよと衝突する場合は右に移動できないかテスト.
    /// </summary>
    [Fact]
    public void ステージのぷよと衝突する場合は右に移動できない()
    {
        // Arrange
        this.stage.Initialize();
        this.stage.SetPuyo(3, 5, 1); // ステージにぷよを配置
        this.player.CreateNewPuyo();
        this.player.SetPuyoX(2);
        this.player.SetPuyoY(5);
        this.player.SetRotation(0); // 上向き

        // Act
        this.player.MoveRight();

        // Assert
        Assert.Equal(2, this.player.PuyoX); // 移動できないので位置は変わらない
    }

    /// <summary>
    /// 子ぷよがステージのぷよと衝突する場合は左に移動できないかテスト.
    /// </summary>
    [Fact]
    public void 子ぷよがステージのぷよと衝突する場合は左に移動できない()
    {
        // Arrange
        this.stage.Initialize();
        this.stage.SetPuyo(2, 5, 1); // ステージにぷよを配置
        this.player.CreateNewPuyo();
        this.player.SetPuyoX(3);
        this.player.SetPuyoY(5);
        this.player.SetRotation(3); // 左向き（子ぷよが x=2 にある）

        // Act
        this.player.MoveLeft();

        // Assert
        Assert.Equal(3, this.player.PuyoX); // 移動できないので位置は変わらない
    }

    /// <summary>
    /// 子ぷよがステージのぷよと衝突する場合は右に移動できないかテスト.
    /// </summary>
    [Fact]
    public void 子ぷよがステージのぷよと衝突する場合は右に移動できない()
    {
        // Arrange
        this.stage.Initialize();
        this.stage.SetPuyo(3, 5, 1); // ステージにぷよを配置
        this.player.CreateNewPuyo();
        this.player.SetPuyoX(2);
        this.player.SetPuyoY(5);
        this.player.SetRotation(1); // 右向き（子ぷよが x=3 にある）

        // Act
        this.player.MoveRight();

        // Assert
        Assert.Equal(2, this.player.PuyoX); // 移動できないので位置は変わらない
    }

    /// <summary>
    /// 右端で右回転した時に左にがべキックされるかテスト.
    /// </summary>
    [Fact]
    public void 右端で右回転した時に左にがべキックされる()
    {
        // Arrange
        this.stage.Initialize();
        this.player.CreateNewPuyo();
        this.player.SetPuyoX(this.config.StageWidth - 1); // 右端に配置
        this.player.SetPuyoY(5);
        this.player.SetRotation(0); // 上向き
        var initialX = this.player.PuyoX;

        // Act
        this.player.RotateRight(); // 右回転すると子ぷよが右端を超える

        // Assert
        Assert.Equal(1, this.player.Rotation); // 回転は成功
        Assert.Equal(initialX - 1, this.player.PuyoX); // 左に1マスずれる
    }

    /// <summary>
    /// 左端で左回転した時に右にがべキックされるかテスト.
    /// </summary>
    [Fact]
    public void 左端で左回転した時に右にがべキックされる()
    {
        // Arrange
        this.stage.Initialize();
        this.player.CreateNewPuyo();
        this.player.SetPuyoX(0); // 左端に配置
        this.player.SetPuyoY(5);
        this.player.SetRotation(0); // 上向き
        var initialX = this.player.PuyoX;

        // Act
        this.player.RotateLeft(); // 左回転すると子ぷよが左端を超える

        // Assert
        Assert.Equal(3, this.player.Rotation); // 回転は成功
        Assert.Equal(initialX + 1, this.player.PuyoX); // 右に1マスずれる
    }

    /// <summary>
    /// 回転時に他のぷよと衝突する場合は回転できないかテスト.
    /// </summary>
    [Fact]
    public void 回転時に他のぷよと衝突する場合は回転できない()
    {
        // Arrange
        this.stage.Initialize();
        this.stage.SetPuyo(3, 5, 1); // 回転先にぷよを配置
        this.player.CreateNewPuyo();
        this.player.SetPuyoX(2);
        this.player.SetPuyoY(5);
        this.player.SetRotation(0); // 上向き
        var initialRotation = this.player.Rotation;

        // Act
        this.player.RotateRight(); // 右回転すると子ぷよが (3, 5) に衝突

        // Assert
        Assert.Equal(initialRotation, this.player.Rotation); // 回転できない
    }
}
